hibernate-core/src/main/java/org/hibernate/loader/ast/internal/EntityBatchLoaderArrayParam.java

```diff
@@ -125,11 +117,13 @@ public T load(
 		}
 	}
 
-	protected Object[] resolveIdsToInitialize(Object pkValue, SharedSessionContractImplementor session) {
-		final Object[] idsToLoad = (Object[]) Array.newInstance( identifierMapping.getJavaType().getJavaTypeClass(), domainBatchSize );
+	@SuppressWarnings( "unchecked" )
+	protected <X> X[] resolveIdsToInitialize(Object pkValue, SharedSessionContractImplementor session) {
+		final X[] idsToLoad = (X[]) Array.newInstance( identifierMapping.getJdbcMapping().getJdbcJavaType().getJavaTypeClass(), domainBatchSize );
+
 		session.getPersistenceContextInternal().getBatchFetchQueue().collectBatchLoadableEntityIds(
 				domainBatchSize,
-				(index, value) -> idsToLoad[index] = value,
+				(index, value) -> idsToLoad[index] = (X) identifierMapping.disassemble( value, session ),
```

Comment:
Shouldn't this rather be this?

```suggestion
				(index, value) -> idsToLoad[index] = (X) identifierMapping.getJdbcMapping().convertToRelationalValue( value ),
```

Reply:
@beikov change applied, thanks!